src/Microsoft.OData.Core/JsonLight/ODataJsonLightReader.cs

```diff
@@ -2256,6 +2259,117 @@ private void EndEntry()
             }
         }
 
+        /// <summary>
+        /// For response de-serialization, restore the null values for properties that are omitted
+        /// by the omit-values=nulls preference header.
+        /// </summary>
+        private void RestoreOmittedNullValues()
+        {
+            // Restore omitted null value only when processing response and when
+            // Preference header omit-values=nulls is specified.
+            if (ShouldRestoreNullValues())
+            {
+                IODataJsonLightReaderResourceState resourceState = this.CurrentResourceState;
+                IEdmStructuredType edmStructuredType = resourceState.ResourceType;
+
+                if (resourceState.SelectedProperties == SelectedPropertiesNode.Empty)
+                {
+                    return;
+                }
+                else
+                {
+                    if (resourceState.Resource != null && edmStructuredType != null)

```

Comment:
>&& edmStructuredType != null [](start = 55, length = 28)

If edmStructuredType is null, why not just treat all properties as dynamic and restore anything that is not in the payload?

Reply:
In what case can edmStructuredType be null?
For OpenType w/ dynamic props, it should be the edm type; for Untype it should be 'IEdmComplexType', right? 

---
In reply to: [198065705](https://github.com/OData/odata.net/pull/1106#discussion_r198065705) [](ancestors = 198065705)

Reply:
then should edmStructuredType != null be an assert, rather than a condition, on line 2287?

---
In reply to: [209787888](https://github.com/OData/odata.net/pull/1106#discussion_r209787888) [](ancestors = 209787888,198065705)

Reply:
sure, updated to Assert(edmStructuredType != null)

---
In reply to: [229534051](https://github.com/OData/odata.net/pull/1106#discussion_r229534051) [](ancestors = 229534051,209787888,198065705)